### PR TITLE
Generate language_id

### LINKED ISSUE #3211: Keep a MAX_LANGUAGE_ID somewhere to prevent language_id dups
As @TwP wrote in https://github.com/github/linguist/pull/3205#issuecomment-246774874, we need to:

> Regarding language ID reuse, keep a MAX_LANGUAGE_ID constant somewhere in the codebase would help ensure that language IDs are not reused in the scenario laid out by @pchaigno. The requirement being that all new language IDs must be larger than MAX_LANGUAGE_ID, and then increment MAX_LANGUAGE_ID after the new language is added.

Adding this here to remind me to do this.

### PR DESCRIPTION
**Pull request by pchaigno**
This pull request implements [@bkeepers' idea](https://github.com/github/linguist/pull/3247#issuecomment-252083158), generating the `language_id` from the hash of the language's name. I also added a test to check that all language ids are unsigned 32bit integers.
Supersedes #3247. Fixes #3211.

The `language_id` is the SHA256 hash of the language name modulo-ed to an unsigned 32bit integer:

``` ruby
Digest::SHA256.hexdigest(language).to_i(16) % (2**30 - 1)
```

Therefore, as long as the language names are different, the probability of collision is of 0.009561175% if we reach 1,000 languages [1]. This is important as we might not always be able to detect a collision. For instance, if a language entry is removed from Linguist and we later add a new language id that conflicts with the older removed language, the tests won't help detecting it.

The choice to generate the language id from the language name is based on the idea that we probably want two entries with the same name to have the same id (if we remove a language and add it back later). If we do not want that, I could simply change the pull request to generate random ids, which would give us the same collision probability.

I think it's important that we discuss this thoroughly, as we probably want to avoid changing the strategy to generate language ids too often (it's already our second system :stuck_out_tongue_closed_eyes: ).
/cc @bkeepers @arfon @Alhadis @TwP 

[1] - ql_15e041ca4f6190549ca56982ef8053e1_l3 - Details at [preshing.com](http://preshing.com/20110504/hash-collision-probabilities/). I took into consideration the fact that we already have 422 fixed language ids.

### CODE REVIEW

File: test/test_language.rb [hunk 1]
```diff
@@ -426,6 +426,14 @@ def test_all_languages_have_a_language_id_set
...
   end
 
+  def test_all_languages_have_a_valid_id
+    invalid = Language.all.select { |language| language.language_id < 0 || language.language_id >= (2**31 - 1) }
+
+    message = "The following languages do not have a valid language_id. Please use script/set-language-ids --update as per the contribution guidelines.\n"
+    invalid.each { |language| message << "#{language.name}\n" }
+    assert invalid.empty?, message
+  end
+
```
Comment (bkeepers) [inline test/test_language.rb]: It'd probably also be worth testing that all the IDs are unique. A collision should be rare, but not impossible.

**Comment by Alhadis** [comment]
Weren't we going to use the digest of the entire `languages.yml` file before the ID is calculated? That's guaranteed to be unique every time, because the hash is being generated based on the entire file's data *before* the `language_id` field is inserted.

For instance, let's assume I'm in the middle of adding a new entry:
```yml
0x00:
  type: programming
  extensions:
  - ".00"
  tm_scope: none
  ace_mode: text
```
Prepending that to the [current state of the file](https://github.com/github/linguist/blob/ffd984bb7e6dec6f82f359308a9a1ea542247191/lib/linguist/languages.yml) yields a checksum of `055d502e8b8a4bd1ea4367772cd1e86e`:

```sh
λ GitHub-Linguist (master): md5 lib/linguist/languages.yml
055d502e8b8a4bd1ea4367772cd1e86e
```

Now, if I add *that* checksum to the entry...

```yml
0x00:
  type: programming
  extensions:
  - ".00"
  tm_scope: none
  ace_mode: text
  language_id: 055d502e8b8a4bd1ea4367772cd1e86e
```

... suddenly, the checksum is different:

```
λ GitHub-Linguist (master): md5 lib/linguist/languages.yml
5f2732b95e6ba0ed1f3d66c97d4be45c
`

Theoretically, that should make it impossible for there to ever be a collision, should it not? The only conflict I can foresee is if two people are adding the same language at the same time, from the same copy of the file. =)

File: test/test_language.rb (hunk 1 above)
Comment (pchaigno) [inline test/test_language.rb]: Unless you're thinking of something different, [there's already a test for that](https://github.com/github/linguist/blob/ffd984bb7e6dec6f82f359308a9a1ea542247191/test/test_language.rb#L429-L435).

**Comment by pchaigno** [comment]
> Weren't we going to use the digest of the entire languages.yml file before the ID is calculated? That's guaranteed to be unique every time

No, it's only guaranteed to be different with a probability that depends on the size of the hash function output (output values are guaranteed to be uniformly distributed). That should be the probability I computed above.
If we do not want the `language_id`s to be the same for two entries with the same language name, then we should simply generate a completely random id. It is simpler and provides us with the same collision probability.

**Comment by Alhadis** [comment]
Well then, perhaps just hashing a concatenation of the name and a random ID?

**Comment by pchaigno** [comment]
@Alhadis That would give us the same result as generating directly a random id ;)

You seem to think that we should generate a different id every time, even if the languages' names are the same. Why is that? Do you see a case where my proposed method would be an issue?

**Comment by Alhadis** [comment]
No, my approach to dealing with numbers I can't grasp is to blast it with as much OTT syrup as possible.

I'll, erm, let you drive... pardon me while I sit in the back seat with all my regular expressions...

**Comment by brandonblack** [comment]
@pchaigno I'm happy with the size / possibility space for generating these hashes. If we start seeing too many collisions we can add some entropy into into how we generate these but I don't think it'll be necessary.

On GitHub's side, this change won't have much impact since this field in our database and search it set for 32-bit integers and this change stays within that. However, I'm not familiar enough with linguist's usage outside for GitHub to really say the same for everyone else.

How much of a disturbance will this change be to others? Are there any special considerations we should think about in releasing this change?

CC: @LizzHale @lildude @TwP

**Comment by pchaigno** [comment]
> If we start seeing too many collisions we can add some entropy into into how we generate these but I don't think it'll be necessary.

Unfortunately there is no way to reduce the probability of collisions other than extending the id bit size since SHA-256 already achieves uniform distribution.

> However, I'm not familiar enough with linguist's usage outside for GitHub to really say the same for everyone else.

I don't know of any major project (outside of github.com) which uses Linguist.

> How much of a disturbance will this change be to others? Are there any special considerations we should think about in releasing this change?

This change only affects the way ids are generated so I don't think there will be any disturbance; it does not affect the API.

**Pull request merged**